tools/check_preset_settings.py

```diff
@@ -0,0 +1,126 @@
+#!/usr/bin/env python
+import configparser
+import json
+import os
+import sys
+
+
+class PresetSettingsValidator:
+
+    def __init__(self, cura_dir: str):
+        self._cura_dir = os.path.abspath(cura_dir)
+        self._resource_dir = os.path.join(self._cura_dir, "resources")
+        self._definitions_dir = os.path.join(self._resource_dir, "definitions")
+        self._preset_settings_dir = os.path.join(self._resource_dir, "preset_setting_visibility_groups")
+
+        self._fdmprinter_def_path = os.path.join(self._definitions_dir, "fdmprinter.def.json")
+
+    def validate(self) -> bool:
+        """
+        Validates the preset settings files and returns True or False indicating whether there are invalid files.
+        """
+        if not os.path.isfile(self._fdmprinter_def_path):
+            raise FileNotFoundError("[%s] is not a file or doesn't exist, please make sure you have specified the correct cura directory [%s]." % (self._fdmprinter_def_path, self._cura_dir))
+
+        if not os.path.isdir(self._preset_settings_dir):
+            raise FileNotFoundError("[%s] is not a directory or doesn't exist, please make sure you have specified the correct cura directory [%s]." % (self._preset_settings_dir, self._cura_dir))
+
+        # parse the definition file
+        setting_tree_dict = self._parse_definition_file(self._fdmprinter_def_path)
+
+        has_invalid_files = False
+
+        # go through all the preset settings files
+        for root_dir, _, filenames in os.walk(self._preset_settings_dir):
+            for filename in filenames:
+                file_path = os.path.join(root_dir, filename)
+                print("Validating [%s] ..." % file_path)
+
+                incorrect_sections = []
+                incorrect_settings = {}
+
+                parser = configparser.ConfigParser(allow_no_value = True)
+                with open(file_path, "r", encoding = "utf-8") as f:
+                    parser.read_file(f)
+
+                for key in parser:
+                    # skip general
+                    if key in ("general", configparser.DEFAULTSECT):
+                        continue
+
+                    if key not in setting_tree_dict:
+                        incorrect_sections.append(key)
+                        continue
+
+                    for setting_key in parser[key]:
+                        if setting_key not in setting_tree_dict[key]:
+                            if setting_key not in incorrect_settings:
+                                incorrect_settings[setting_key] = {"seen_in": [],
+                                                                   "should_be_in": self._should_setting_be_in(setting_tree_dict, setting_key)}
+
+                            incorrect_settings[setting_key]["seen_in"].append(key)
+
+                # show results
+                print("==========================================")
```

Comment:
debug statement?

Reply:
This is an independent script which validates those files and outputs a report in console. That's why it's all prints and not using any cura or uranium modules.

Reply:
I plan to put this script in CI some time later so it will become one of the checks we do on commits. Other checks/scripts can also be included in the future.

Reply:
Ok, my bad.